Stability: Improve Developer Documentation (#8517)
solo-io/gloo: devel/testing/writing-tests.md

# Writing Tests
- [Convention](#conventions)
- [Matchers](#matchers)
  - [Gomega Matchers](#gomega-matchers)
  - [Custom Matchers](#custom-matchers)
- [Transforms](#transforms)
  - [Custom Transforms](#custom-transforms)
- [Assertions](#assertions)
  - [Prefer Explicit Error Checking](#prefer-explicit-error-checking)
  - [Prefer Assertion Descriptions](#prefer-assertion-descriptions)
  - [Prefer Http Response Matcher](#prefer-http-response-matcher)

## Conventions

> - All new packages and most new significant functionality must come with unit tests.
> - Table-driven tests are preferred for testing multiple scenarios/inputs.

Reviewer: ```suggestion
- All new packages and most new significant functionality must come with unit tests
- Table-driven tests are preferred for testing multiple scenarios/inputs
```

- Significant features should come with [end-to-end (test/e2e) tests](e2e-tests.md) and/or [kubernetes end-to-end (test/kube2e) tests](kube-e2e-tests.md)

> - Tests which are platform dependent, should be marked as such using [test requirements](/test/testutils/requirements.go)

Reviewer: ```suggestion
- Tests which are platform-dependent, should be marked as such using [test requirements](/test/testutils/requirements.go)
```

## Matchers
### Gomega Matchers
Gomega has a powerful set of built-in matchers. We recommend using these matchers whenever possible. You can find the full list of matchers [here](https://github.com/onsi/gomega/tree/master/matchers).

### Custom Matchers

> We have a few custom matchers that we use in our tests. These are defined in a [matchers package](/test/gomega/matchers/). If you find yourself writing a custom matcher, consider adding it to this package.

Reviewer: ```suggestion
We have a number of custom matchers that we use in our tests. These are defined in a [matchers package](/test/gomega/matchers/). If you find yourself writing a custom matcher, consider adding it to this package.
```
At some point "few" may no longer be an accurate description

## Transforms
It is possible to [compose matchers using transforms](https://onsi.github.io/gomega/#composing-matchers). Transforms are either:
- functions which accept one parameter that returns one value

> - functions which accept one parameter that returns two values, where the second value must be of the error type.

Reviewer: ```suggestion
- functions which accept one parameter that returns two values, where the second value must be of the error type
```

Transforms allow us to re-use matchers, and convert the data that we want to compare into a format that the matcher can understand. Let's say we want to compare the data returned by an http.Response to a key/value pair:
```go

> Expect(response).To(HaveKeyWithValue("queryStringParameters", HaveKeyWithValue("foo", "bar")))

Reviewer: I'm a bit confused about what's going on with the nested `HaveKeyWithValue` matchers
Could we clarify or use a simpler example?

Reviewer: The broken example has nested calls to `HaveKeyWithValue()`
Why/how would `HaveKeyWithValue("foo", "bar")` be a viable second argument to `HaveKeyWithValue()`?
This is the piece that's confusing, and I don't think it's really key to the point that's being made about being able to use a transform to make certain input consumable by matchers that do not normally work for that input

Would it get the same point across to have an example that only goes one layer deep, such as:
```suggestion
Expect(response).To(HaveKeyWithValue("status", "200"))
```

and then 
```go
Expect(response).To(WithTransform(transforms.WithJsonBody(), HaveKeyWithValue("status", "200")))
```

as the fixed example

Author: Good point! I added another example just to drive the point home, but also updated the http example to use the case you provided. I like it too because it avoids foo/bar and instead deals with a real use case

```

This doesn't work, because the response (*http.Response) is not a map[string]interface{}, so we can't use the standard `HaveKeyWithValue` matcher. We can use a transform to convert the response into a map[string]interface{}:
```go
WithTransform(transforms.WithJsonBody(), {MATCHER})
```

Now we can rewrite our assertion as:
```go
Expect(response).To(WithTransform(transforms.WithJsonBody(), HaveKeyWithValue("queryStringParameters", HaveKeyWithValue("foo", "bar"))))
````

### Custom Transforms
We have a few custom matchers that we use in our tests. These are defined in a [transforms package](/test/gomega/transforms/). If you find yourself writing a custom transform, consider adding it to this package.

## Assertions
### Prefer Explicit Error Checking
A common pattern to assert than an error occurred
```go
Expect(err).To(HaveOccurred())
```

A more explict way to perfrom this assertion is:
```go
Expect(err).To(MatchError("expected error"))
```

> 

Reviewer: ````suggestion
or
```go
Expect(err).To(MatchError(GlobalError))
```
or
```go
Expect(err).To(MatchError(ErrorFunc("expected error"))
```

````
making the strictest assertion possible

## Transforms
It is possible to [compose matchers using transforms](https://onsi.github.io/gomega/#composing-matchers). Transforms are either:
- functions which accept one parameter that returns one value
- functions which accept one parameter that returns two values, where the second value must be of the error type

Transforms allow us to re-use matchers, and convert the data that we want to compare into a format that the matcher can understand. Let's say we want to compare the data returned by an http.Response to a key/value pair:
```go
Expect(response).To(HaveKeyWithValue("queryStringParameters", HaveKeyWithValue("foo", "bar")))
```

This doesn't work, because the response (*http.Response) is not a map[string]interface{}, so we can't use the standard `HaveKeyWithValue` matcher. We can use a transform to convert the response into a map[string]interface{}:
```go
WithTransform(transforms.WithJsonBody(), {MATCHER})
```

Now we can rewrite our assertion as:
```go
Expect(response).To(WithTransform(transforms.WithJsonBody(), HaveKeyWithValue("queryStringParameters", HaveKeyWithValue("foo", "bar"))))

> ````

Reviewer: ````suggestion
```
````

- functions which accept one parameter that returns one value
- functions which accept one parameter that returns two values, where the second value must be of the error type

> Transforms allow us to re-use matchers, and convert the data that we want to compare into a format that the matcher can understand. Below are a few example that illustrate this concept.

Reviewer: ```suggestion
Transforms allow us to re-use matchers, and convert the data that we want to compare into a format that the matcher can understand. Below are a couple examples that illustrate this concept.
```
